#### Version of the Library:

In [1]:
#python version
import sys
print('Python :{}'.format(sys.version))

#scipy
import scipy
print('Scipy :{}'.format(scipy.__version__))

#Numpy
import numpy
print('Numpy :{}'.format(numpy.__version__))

#Matplotlib
import matplotlib
print('Matplotlib :{}'.format(matplotlib.__version__))

#Pandas
import pandas
print('Pandas :{}'.format(pandas.__version__))

#Sklearn
import sklearn
print('Sklearn :{}'.format(pandas.__version__))

#Flask
import flask
print('Flask :{}'.format(flask.__version__))

import conda
print('conda :{}'.format(conda.__version__))

Python :3.7.6 (default, Jan  8 2020, 20:23:39) [MSC v.1916 64 bit (AMD64)]
Scipy :1.5.2
Numpy :1.19.1
Matplotlib :3.3.1
Pandas :1.1.4
Sklearn :1.1.4
Flask :1.1.2
conda :4.9.2


#### Import Library

In [2]:
#Loading all the libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

#### Load data: as csv file

In [3]:
recipes = pd.read_csv("Recipes.csv")
defects = pd.read_csv("Defects.csv")

In [4]:
recipes.head(3)

HEAT_ID  Material_Code            NAME    WEIGHT
0  1701190            112  # 2 HEAVY MELT  158758.0
1  1701192            113        PIG IRON   50306.6
2  1701192            115      HOME SCRAP    2215.8

In [5]:
recipes.describe()

HEAT_ID  Material_Code         WEIGHT
count  1.000020e+05  100002.000000  100002.000000
mean   1.749167e+06     129.451171   19169.098051
std    4.765453e+04      24.834632   17224.494762
min    1.701190e+06     112.000000       0.500000
25%    1.707906e+06     115.000000    4822.100000
50%    1.715312e+06     120.000000   10813.900000
75%    1.804858e+06     126.000000   30399.675000
max    1.811206e+06     215.000000  158758.000000

In [6]:
defects.head(3)

PIECE_ID  HEAT_ID  DEFECT_TYPE DEFECT_NAME DEFECT_GROUP_ID  \
0    167872  1701192           59       Scabs              SU   
1    167867  1701192           59       Scabs              SU   
2    167869  1701192           63   Scratches              SU   

  DEFECT_GROUP_Name  
0         1-Surface  
1         1-Surface  
2         1-Surface

In [7]:
defects.describe()

PIECE_ID       HEAT_ID    DEFECT_TYPE
count  563048.000000  5.630480e+05  563048.000000
mean   298638.715351  1.744834e+06     175.309350
std     83246.354402  4.680686e+04      33.533375
min    166934.000000  1.701192e+06      36.000000
25%    224145.000000  1.706120e+06     167.000000
50%    299206.000000  1.714372e+06     182.000000
75%    358950.250000  1.802874e+06     190.000000
max    679818.000000  1.811206e+06     238.000000

### Data profiling using Pandas:

In [8]:
import pandas_profiling as pp
#pp.ProfileReport(recipes)

In [9]:
import pandas_profiling as pp
#pp.ProfileReport(defects)

In [10]:
#defects.describe()

In [11]:
print(f"Recipes data: \n")
print(recipes.head(2))
print("----------------------------------------------------------------------")
print(f"Defects data:\n")
print(defects.head(2))

Recipes data: 

   HEAT_ID  Material_Code            NAME    WEIGHT
0  1701190            112  # 2 HEAVY MELT  158758.0
1  1701192            113        PIG IRON   50306.6
----------------------------------------------------------------------
Defects data:

   PIECE_ID  HEAT_ID  DEFECT_TYPE DEFECT_NAME DEFECT_GROUP_ID  \
0    167872  1701192           59       Scabs              SU   
1    167867  1701192           59       Scabs              SU   

  DEFECT_GROUP_Name  
0         1-Surface  
1         1-Surface  


# Data preprocessing or Data Cleaning:

###### data Info & Unique value:

In [15]:
#uniq name from data
print(f"Unique HEAT_ID of Recipes: {recipes['Material_Code'].unique()}\n")
print(f"Unique HEAT_ID of Recipes: {recipes['HEAT_ID'].unique()}")
print("------------------------------------------------------------------------------------")
print(f"Unique DEFECT_TYPE of defects: {defects['DEFECT_TYPE'].unique()}")
print(f"Unique DEFECT_Group_Id of defects: {defects['DEFECT_GROUP_ID'].unique()}")

print(len(recipes['Material_Code'].unique()))

print(len(defects['DEFECT_TYPE'].unique()))

Unique HEAT_ID of Recipes: [112 113 115 114 117 119 121 126 124 120 116 180 166 183 182 190 181 122
 125 118 208 209 207 210 211 215 212 214 213]

Unique HEAT_ID of Recipes: [1701190 1701192 1701194 ... 1811202 1811204 1811206]
------------------------------------------------------------------------------------
Unique DEFECT_TYPE of defects: [ 59  63  71  89  90 146 148 149 150 167 168 169 182 190 151 157 147 197
 134 153 155 154 156 211 193  56  46  87  49  60  52 212 105 186 130 118
 104 125  96 173 165  88 103 114 175 176 100  99  41  66 126 191 192  91
  55 184  53 121  79 204 187 185 194 107  94 102  47 214  86  68 101 198
  43  80 183 174 110 119  61 188  74 120 109  92 170 133  54  95 111  62
  78 108 113  37  65  39 213  73  44 128 201 115  98 208  75  97  69  81
 202 135  38 152 207 199  77 136 216 196  36 205 117  72  82  93  45  76
  42 164  64 210 129 127  57 200 116  83 222 162 223 195 171 122 206 158
  58 178 177 172  48  50 209 215 106  67  51  40 238 217 112 159 160  84

In [13]:
print(f"Length of Recipes data: {len(recipes)} and Shape of the data {recipes.shape}")
print(f"Length of Defects data: {len(defects)} and Shape of the data {defects.shape}")

Length of Recipes data: 100002 and Shape of the data (100002, 4)
Length of Defects data: 563048 and Shape of the data (563048, 6)


In [14]:
print(f"All Columns name of Recipes: {list(recipes.columns)}")
print(f"All Columns name of Defects: {list(defects.columns)}")

All Columns name of Recipes: ['HEAT_ID', 'Material_Code', 'NAME', 'WEIGHT']
All Columns name of Defects: ['PIECE_ID', 'HEAT_ID', 'DEFECT_TYPE', 'DEFECT_NAME', 'DEFECT_GROUP_ID', 'DEFECT_GROUP_Name']


In [15]:
#Make Pivot table for 'recipes.csv' file base on 'HEAT_ID'
df = pd.pivot_table(recipes, index="HEAT_ID", columns=['Material_Code'], values='WEIGHT')

df.reset_index(inplace=True)



In [16]:
df.head(2)

Material_Code  HEAT_ID       112      113     114     115  116  117  118  119  \
0              1701190  158758.0      NaN     NaN     NaN  NaN  NaN  NaN  NaN   
1              1701192       NaN  50306.6  3473.6  2215.8  NaN  NaN  NaN  NaN   

Material_Code  120  ...  190  207  208  209  210  211  212  213  214  215  
0              NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
1              NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  

[2 rows x 30 columns]

###### Add New Columns:Num,Sum,Mean,Repeat

In [17]:
#Add Sum Columns
df["Num"]=df.iloc[:,1:30].count(axis=1)
df['Sum'] = df.iloc[:, 1:30].sum(axis=1)
df['Mean'] = df.iloc[:, 1:30].mean(axis=1)



In [18]:
df

Material_Code  HEAT_ID       112      113     114     115     116      117  \
0              1701190  158758.0      NaN     NaN     NaN     NaN      NaN   
1              1701192       NaN  50306.6  3473.6  2215.8     NaN      NaN   
2              1701194       NaN  49903.8  3426.0  2322.8     NaN  34129.7   
3              1701196       NaN  49893.9  3173.8  1812.6  3173.8  35378.9   
4              1701198       NaN  22624.8  1544.9  1046.0     NaN  15114.6   
...                ...       ...      ...     ...     ...     ...      ...   
12741          1811198       NaN  31282.9     NaN     NaN     NaN  50207.3   
12742          1811200       NaN  32899.5     NaN     NaN     NaN  50506.7   
12743          1811202       NaN  31732.0     NaN     NaN     NaN  51175.3   
12744          1811204       NaN  32145.2     NaN     NaN     NaN  51277.3   
12745          1811206       NaN  32147.5     NaN     NaN     NaN  51038.7   

Material_Code     118     119     120  ...  209  210  211  212     213  214  \
0                 NaN     NaN     NaN  ...  NaN  NaN  NaN  NaN     NaN  NaN   
1                 NaN     NaN     NaN  ...  NaN  NaN  NaN  NaN     NaN  NaN   
2                 NaN  4545.9  2180.0  ...  NaN  NaN  NaN  NaN     NaN  NaN   
3                 NaN  4533.7  1812.6  ...  NaN  NaN  NaN  NaN     NaN  NaN   
4                 NaN  2066.6  1106.8  ...  NaN  NaN  NaN  NaN     NaN  NaN   
...               ...     ...     ...  ...  ...  ...  ...  ...     ...  ...   
12741          6172.5  6422.9     NaN  ...  NaN  NaN  NaN  NaN  8431.4  NaN   
12742          7185.4  6217.4     NaN  ...  NaN  NaN  NaN  NaN  7839.4  NaN   
12743          6381.1  6274.6     NaN  ...  NaN  NaN  NaN  NaN  8656.4  NaN   
12744          6783.5  7053.8     NaN  ...  NaN  NaN  NaN  NaN  8866.8  NaN   
12745          6841.5  6534.9     NaN  ...  NaN  NaN  NaN  NaN  8785.2  NaN   

Material_Code  215  Num       Sum           Mean  
0              NaN    1  158758.0  158758.000000  
1              NaN    3   55996.0   18665.333333  
2              NaN    9  160919.8   17879.977778  
3              NaN   10  139237.8   13923.780000  
4              NaN    8   50799.3    6349.912500  
...            ...  ...       ...            ...  
12741          NaN    9  151817.6   16868.622222  
12742          NaN    9  152339.2   16926.577778  
12743          NaN    9  152226.8   16914.088889  
12744          NaN    9  152497.8   16944.200000  
12745          NaN    9  152180.4   16908.933333  

[12746 rows x 33 columns]

In [19]:
print(f"Shape of Recipies the data: {df.shape}")

Shape of Recipies the data: (12746, 33)


#### Merge Data set:

In [20]:
#Marge recipes data with defect data,Fill with 0, Drop the duplicates values
df = pd.merge(defects, df, on = 'HEAT_ID')
df = df.fillna(0)

#Drop Duplicates
df=df.drop_duplicates()

#Drop duplicate columns
df = df.loc[:,~df.columns.duplicated()]
df


df.shape

(419277, 38)

In [21]:
df.head(4)

PIECE_ID  HEAT_ID  DEFECT_TYPE DEFECT_NAME DEFECT_GROUP_ID  \
0    167872  1701192           59       Scabs              SU   
1    167867  1701192           59       Scabs              SU   
2    167869  1701192           63   Scratches              SU   
3    167871  1701192           71       Stain              SU   

  DEFECT_GROUP_Name  112      113     114     115  ...  209  210  211  212  \
0         1-Surface  0.0  50306.6  3473.6  2215.8  ...  0.0  0.0  0.0  0.0   
1         1-Surface  0.0  50306.6  3473.6  2215.8  ...  0.0  0.0  0.0  0.0   
2         1-Surface  0.0  50306.6  3473.6  2215.8  ...  0.0  0.0  0.0  0.0   
3         1-Surface  0.0  50306.6  3473.6  2215.8  ...  0.0  0.0  0.0  0.0   

   213  214  215  Num      Sum          Mean  
0  0.0  0.0  0.0    3  55996.0  18665.333333  
1  0.0  0.0  0.0    3  55996.0  18665.333333  
2  0.0  0.0  0.0    3  55996.0  18665.333333  
3  0.0  0.0  0.0    3  55996.0  18665.333333  

[4 rows x 38 columns]

In [22]:
#Sorting/Describing Data
#df.sort_values(['DEFECT_TYPE', 'DEFECT_GROUP_ID'], ascending=[0,1])
#df.head(4)

In [23]:
#Check the Null Value
#df.isnull().sum()

In [24]:
#Add Repeat Columns

df['Repeat']=df.groupby('HEAT_ID').HEAT_ID.transform('count')


In [25]:
df

PIECE_ID  HEAT_ID  DEFECT_TYPE                           DEFECT_NAME  \
0         167872  1701192           59                                 Scabs   
1         167867  1701192           59                                 Scabs   
2         167869  1701192           63                             Scratches   
3         167871  1701192           71                                 Stain   
4         167868  1701192           71                                 Stain   
...          ...      ...          ...                                   ...   
511067    477203  1811206          182                   pending PQA results   
511068    477203  1811206          212                  Pending Test Results   
511069    477203  1811206          193   Length Questionable - Verify Length   
511070    477203  1811206          190  Wait For Weight - No L2 Scale Weight   
511071    477203  1811206          216   Bad Dimensions on Calculated Weight   

       DEFECT_GROUP_ID DEFECT_GROUP_Name  112      113     114     115  ...  \
0                   SU         1-Surface  0.0  50306.6  3473.6  2215.8  ...   
1                   SU         1-Surface  0.0  50306.6  3473.6  2215.8  ...   
2                   SU         1-Surface  0.0  50306.6  3473.6  2215.8  ...   
3                   SU         1-Surface  0.0  50306.6  3473.6  2215.8  ...   
4                   SU         1-Surface  0.0  50306.6  3473.6  2215.8  ...   
...                ...               ...  ...      ...     ...     ...  ...   
511067             OTH           0-Other  0.0  32147.5     0.0     0.0  ...   
511068             OTH           0-Other  0.0  32147.5     0.0     0.0  ...   
511069              DM       3-Dimension  0.0  32147.5     0.0     0.0  ...   
511070              DM       3-Dimension  0.0  32147.5     0.0     0.0  ...   
511071             OTH           0-Other  0.0  32147.5     0.0     0.0  ...   

        210  211  212     213  214  215  Num       Sum          Mean  Repeat  
0       0.0  0.0  0.0     0.0  0.0  0.0    3   55996.0  18665.333333      79  
1       0.0  0.0  0.0     0.0  0.0  0.0    3   55996.0  18665.333333      79  
2       0.0  0.0  0.0     0.0  0.0  0.0    3   55996.0  18665.333333      79  
3       0.0  0.0  0.0     0.0  0.0  0.0    3   55996.0  18665.333333      79  
4       0.0  0.0  0.0     0.0  0.0  0.0    3   55996.0  18665.333333      79  
...     ...  ...  ...     ...  ...  ...  ...       ...           ...     ...  
511067  0.0  0.0  0.0  8785.2  0.0  0.0    9  152180.4  16908.933333     122  
511068  0.0  0.0  0.0  8785.2  0.0  0.0    9  152180.4  16908.933333     122  
511069  0.0  0.0  0.0  8785.2  0.0  0.0    9  152180.4  16908.933333     122  
511070  0.0  0.0  0.0  8785.2  0.0  0.0    9  152180.4  16908.933333     122  
511071  0.0  0.0  0.0  8785.2  0.0  0.0    9  152180.4  16908.933333     122  

[419277 rows x 39 columns]

In [26]:
#Create CSV file for dataset
#df.to_csv('test.csv')

One Hot Encoding for DEFECT_GROUP_ID:

In [27]:
print(f"Unique HEAT_ID of Recipes: {df['DEFECT_GROUP_ID'].unique()}")

Unique HEAT_ID of Recipes: ['SU' 'SH' 'DM' 'OTH' 'MET' 'CCTO' 'IDF' 'PR' 'PK' 'EQ']


In [28]:
#dummies = pd.get_dummies(df.DEFECT_GROUP_ID)
#dummies.head(3)

#dummies = pd.get_dummies(df.DEFECT_TYPE)
#dummies.head(3)

In [29]:
#concate 
#df = pd.concat([df,dummies.drop(36,axis='columns')],axis='columns')
#df.head()

In [30]:
df.columns

Index([         'PIECE_ID',           'HEAT_ID',       'DEFECT_TYPE',
             'DEFECT_NAME',   'DEFECT_GROUP_ID', 'DEFECT_GROUP_Name',
                       112,                 113,                 114,
                       115,                 116,                 117,
                       118,                 119,                 120,
                       121,                 122,                 124,
                       125,                 126,                 166,
                       180,                 181,                 182,
                       183,                 190,                 207,
                       208,                 209,                 210,
                       211,                 212,                 213,
                       214,                 215,               'Num',
                     'Sum',              'Mean',            'Repeat'],
      dtype='object')

Drop Columns:

In [31]:
df = df.drop(columns=['PIECE_ID','DEFECT_TYPE', 'DEFECT_NAME', 'DEFECT_GROUP_Name','Num','Mean','Repeat'])
df.reset_index(inplace=True)
df

index  HEAT_ID DEFECT_GROUP_ID  112      113     114     115  116  \
0            0  1701192              SU  0.0  50306.6  3473.6  2215.8  0.0   
1            1  1701192              SU  0.0  50306.6  3473.6  2215.8  0.0   
2            2  1701192              SU  0.0  50306.6  3473.6  2215.8  0.0   
3            3  1701192              SU  0.0  50306.6  3473.6  2215.8  0.0   
4            4  1701192              SU  0.0  50306.6  3473.6  2215.8  0.0   
...        ...      ...             ...  ...      ...     ...     ...  ...   
419272  511067  1811206             OTH  0.0  32147.5     0.0     0.0  0.0   
419273  511068  1811206             OTH  0.0  32147.5     0.0     0.0  0.0   
419274  511069  1811206              DM  0.0  32147.5     0.0     0.0  0.0   
419275  511070  1811206              DM  0.0  32147.5     0.0     0.0  0.0   
419276  511071  1811206             OTH  0.0  32147.5     0.0     0.0  0.0   

            117     118  ...  207  208  209  210  211  212     213  214  215  \
0           0.0     0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0   
1           0.0     0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0   
2           0.0     0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0   
3           0.0     0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0   
4           0.0     0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0   
...         ...     ...  ...  ...  ...  ...  ...  ...  ...     ...  ...  ...   
419272  51038.7  6841.5  ...  0.0  0.0  0.0  0.0  0.0  0.0  8785.2  0.0  0.0   
419273  51038.7  6841.5  ...  0.0  0.0  0.0  0.0  0.0  0.0  8785.2  0.0  0.0   
419274  51038.7  6841.5  ...  0.0  0.0  0.0  0.0  0.0  0.0  8785.2  0.0  0.0   
419275  51038.7  6841.5  ...  0.0  0.0  0.0  0.0  0.0  0.0  8785.2  0.0  0.0   
419276  51038.7  6841.5  ...  0.0  0.0  0.0  0.0  0.0  0.0  8785.2  0.0  0.0   

             Sum  
0        55996.0  
1        55996.0  
2        55996.0  
3        55996.0  
4        55996.0  
...          ...  
419272  152180.4  
419273  152180.4  
419274  152180.4  
419275  152180.4  
419276  152180.4  

[419277 rows x 33 columns]

#### Final Data set:

In [32]:
df2=df[['HEAT_ID',112,113,114,115,116,117,118,119,120,121,122,124,125,126,166,180,181,182,183,190,207,208,209,210,211,212,213,214,215,'Sum','DEFECT_GROUP_ID']]
df2


HEAT_ID  112      113     114     115  116      117     118     119  \
0       1701192  0.0  50306.6  3473.6  2215.8  0.0      0.0     0.0     0.0   
1       1701192  0.0  50306.6  3473.6  2215.8  0.0      0.0     0.0     0.0   
2       1701192  0.0  50306.6  3473.6  2215.8  0.0      0.0     0.0     0.0   
3       1701192  0.0  50306.6  3473.6  2215.8  0.0      0.0     0.0     0.0   
4       1701192  0.0  50306.6  3473.6  2215.8  0.0      0.0     0.0     0.0   
...         ...  ...      ...     ...     ...  ...      ...     ...     ...   
419272  1811206  0.0  32147.5     0.0     0.0  0.0  51038.7  6841.5  6534.9   
419273  1811206  0.0  32147.5     0.0     0.0  0.0  51038.7  6841.5  6534.9   
419274  1811206  0.0  32147.5     0.0     0.0  0.0  51038.7  6841.5  6534.9   
419275  1811206  0.0  32147.5     0.0     0.0  0.0  51038.7  6841.5  6534.9   
419276  1811206  0.0  32147.5     0.0     0.0  0.0  51038.7  6841.5  6534.9   

        120  ...  208  209  210  211  212     213  214  215       Sum  \
0       0.0  ...  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0   55996.0   
1       0.0  ...  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0   55996.0   
2       0.0  ...  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0   55996.0   
3       0.0  ...  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0   55996.0   
4       0.0  ...  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0   55996.0   
...     ...  ...  ...  ...  ...  ...  ...     ...  ...  ...       ...   
419272  0.0  ...  0.0  0.0  0.0  0.0  0.0  8785.2  0.0  0.0  152180.4   
419273  0.0  ...  0.0  0.0  0.0  0.0  0.0  8785.2  0.0  0.0  152180.4   
419274  0.0  ...  0.0  0.0  0.0  0.0  0.0  8785.2  0.0  0.0  152180.4   
419275  0.0  ...  0.0  0.0  0.0  0.0  0.0  8785.2  0.0  0.0  152180.4   
419276  0.0  ...  0.0  0.0  0.0  0.0  0.0  8785.2  0.0  0.0  152180.4   

        DEFECT_GROUP_ID  
0                    SU  
1                    SU  
2                    SU  
3                    SU  
4                    SU  
...                 ...  
419272              OTH  
419273              OTH  
419274               DM  
419275               DM  
419276              OTH  

[419277 rows x 32 columns]

In [33]:
df2.shape

(419277, 32)

In [34]:
#Make area(SqFt) vs price plot
#%matplotlib inline
#Make label area and price
#plt.xlabel('HEAT_ID')
#plt.ylabel('DEFECT_GROUP_ID')
#plt.scatter(df2.HEAT_ID,df.DEFECT_GROUP_ID, color ='blue', marker='.') #ekhane kon columns ar plot hobe

In [35]:
#df3 = df2.drop_duplicates()
#df3

In [36]:
#df3.shape

In [37]:
import pandas_profiling as pp
#pp.ProfileReport(df3)

In [38]:
#df3.columns

## Select Model and Tharin the Dataset:

Split The Data X and y :

In [39]:
#importing the data_process file which includes data preprocessing functions
from data_preprocess import*

#Loading all the libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.metrics import precision_score, recall_score, f1_score
from yellowbrick.classifier import ClassificationReport
from sklearn.linear_model import LogisticRegression
from matplotlib.colors import ListedColormap
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.naive_bayes import GaussianNB

In [40]:
X = df2.drop(['DEFECT_GROUP_ID'],axis='columns')
X.head(3)

HEAT_ID  112      113     114     115  116  117  118  119  120  ...  207  \
0  1701192  0.0  50306.6  3473.6  2215.8  0.0  0.0  0.0  0.0  0.0  ...  0.0   
1  1701192  0.0  50306.6  3473.6  2215.8  0.0  0.0  0.0  0.0  0.0  ...  0.0   
2  1701192  0.0  50306.6  3473.6  2215.8  0.0  0.0  0.0  0.0  0.0  ...  0.0   

   208  209  210  211  212  213  214  215      Sum  
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  55996.0  
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  55996.0  
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  55996.0  

[3 rows x 31 columns]

In [41]:
y = df2.DEFECT_GROUP_ID

y.head(3)

0    SU
1    SU
2    SU
Name: DEFECT_GROUP_ID, dtype: object

In [42]:
from sklearn.model_selection import train_test_split
#split Train and test X,y data
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2,random_state=10)

In [43]:
print(f"Length of X_train :{len(X_train)}")
print(f"Length of y_train :{len(y_train)}")
print(f"Length of X_test :{len(X_test)}")
print(f"Length of y_test :{len(y_test)}")

Length of X_train :335421
Length of y_train :335421
Length of X_test :83856
Length of y_test :83856


In [44]:
X

HEAT_ID  112      113     114     115  116      117     118     119  \
0       1701192  0.0  50306.6  3473.6  2215.8  0.0      0.0     0.0     0.0   
1       1701192  0.0  50306.6  3473.6  2215.8  0.0      0.0     0.0     0.0   
2       1701192  0.0  50306.6  3473.6  2215.8  0.0      0.0     0.0     0.0   
3       1701192  0.0  50306.6  3473.6  2215.8  0.0      0.0     0.0     0.0   
4       1701192  0.0  50306.6  3473.6  2215.8  0.0      0.0     0.0     0.0   
...         ...  ...      ...     ...     ...  ...      ...     ...     ...   
419272  1811206  0.0  32147.5     0.0     0.0  0.0  51038.7  6841.5  6534.9   
419273  1811206  0.0  32147.5     0.0     0.0  0.0  51038.7  6841.5  6534.9   
419274  1811206  0.0  32147.5     0.0     0.0  0.0  51038.7  6841.5  6534.9   
419275  1811206  0.0  32147.5     0.0     0.0  0.0  51038.7  6841.5  6534.9   
419276  1811206  0.0  32147.5     0.0     0.0  0.0  51038.7  6841.5  6534.9   

        120  ...  207  208  209  210  211  212     213  214  215       Sum  
0       0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0   55996.0  
1       0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0   55996.0  
2       0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0   55996.0  
3       0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0   55996.0  
4       0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0   55996.0  
...     ...  ...  ...  ...  ...  ...  ...  ...     ...  ...  ...       ...  
419272  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  8785.2  0.0  0.0  152180.4  
419273  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  8785.2  0.0  0.0  152180.4  
419274  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  8785.2  0.0  0.0  152180.4  
419275  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  8785.2  0.0  0.0  152180.4  
419276  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  8785.2  0.0  0.0  152180.4  

[419277 rows x 31 columns]

In [45]:
y

0          SU
1          SU
2          SU
3          SU
4          SU
         ... 
419272    OTH
419273    OTH
419274     DM
419275     DM
419276    OTH
Name: DEFECT_GROUP_ID, Length: 419277, dtype: object

In [48]:
#from sklearn.linear_model import LinearRegression
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression(solver='liblinear',multi_class='ovr' ,C=10)
lr.fit(X_train, y_train)
lr.score(X_test, y_test)


0.4645105895821408

In [49]:
lr.predict([[1701194, 0.0, 0.0, 0.0, 0.0, 0.0,0.0, 0.0, 4545.9, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 4545.9]])

array(['OTH'], dtype=object)

In [60]:
from sklearn.neighbors import KNeighborsClassifier

model = KNeighborsClassifier(n_neighbors=3)

# Train the model using the training sets
model.fit(X,y)

#Predict Output
model.score(X_test,y_test)

0.4365221331806907

In [61]:
import sklearn.metrics as metrics
from sklearn.metrics import f1_score
from sklearn.metrics import recall_score
from sklearn.metrics import precision_score
prediction = model.predict(X_test)

print(f"hamming_loss : {metrics.hamming_loss(y_test, prediction)}")
print(f"f1_score : {f1_score(y_test, prediction, average='macro')}")
print(f"recall_score : {recall_score(y_test, prediction, average='micro')}")
print(f"precision_score : {precision_score(y_test, prediction, average='micro')}")



hamming_loss : 0.5634778668193093
f1_score : 0.1604995232317231
recall_score : 0.4365221331806907
precision_score : 0.4365221331806907


# Ensemble Learning 

## 1. Simple Ensemble Techniques

#### Max Voting : “VotingClassifier” module in sklearn as follows:

In [76]:

from sklearn.ensemble import VotingClassifier
model1 = LogisticRegression(random_state=1)
model2 = DecisionTreeClassifier(random_state=1)
model3 = KNeighborsClassifier()
model4 = RandomForestClassifier(max_depth=10, n_estimators=40)
model = VotingClassifier(estimators=[('lr', model1), ('dt', model2),('kn',model3),('RF',model4)], voting='hard')
model.fit(X_train,y_train)
model.score(X_test,y_test)


C:\Users\rahim\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


0.464605991223049

In [77]:
model.predict([[1701194, 0.0, 0.0, 0.0, 0.0, 0.0,0.0, 0.0, 4545.9, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 4545.9]])

array(['DM'], dtype=object)

In [48]:
'''
from sklearn.ensemble import VotingClassifier
model1 = RandomForestClassifier(max_depth=10, n_estimators=40)
model2 = DecisionTreeClassifier(random_state=1)
model = VotingClassifier(estimators=[('RF', model1), ('dt', model2)], voting='hard')
model.fit(X_train,y_train)
model.score(X_test,y_test)
'''

"\nfrom sklearn.ensemble import VotingClassifier\nmodel1 = RandomForestClassifier(max_depth=10, n_estimators=40)\nmodel2 = DecisionTreeClassifier(random_state=1)\nmodel = VotingClassifier(estimators=[('RF', model1), ('dt', model2)], voting='hard')\nmodel.fit(X_train,y_train)\nmodel.score(X_test,y_test)\n"

In [49]:
#model.predict([[1701194, 0.0, 0.0, 0.0, 0.0, 0.0,0.0, 0.0, 4545.9, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 4545.9]])

## Advanced Ensemble Techniques:: Bagging

In [50]:
'''
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, ExtraTreesClassifier, BaggingClassifier, VotingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.model_selection import cross_val_score

lr_clf = LogisticRegression() 
dt_clf = DecisionTreeClassifier() 
rf_clf = RandomForestClassifier() 
et_clf = ExtraTreesClassifier() 
knn_clf = KNeighborsClassifier() 

classifiers =[lr_clf, dt_clf, rf_clf, et_clf, knn_clf]
for clf in classifiers:
    clf_scores = cross_val_score(clf, X_train, y_train, cv = 10)
    bagging_clf = BaggingClassifier(clf, max_samples=0.4, max_features=10, random_state=0)
    bagging_clf_scores = cross_val_score(bagging_clf, X_train, y_train, cv = 10)
    print(clf.__class__.__name__, ":::: Mean:", clf_scores.mean(), ", Std Dev:", clf_scores.std())
    print("Bagging", clf.__class__.__name__, ":::: Mean:", bagging_clf_scores.mean(), "Std Dev:", bagging_clf_scores.std(), "\n")

ensembler = VotingClassifier(estimators=[('LogisticRegression', lr_clf), ('DecisionTreeClassifier', dt_clf),
                                        ('RandomForestClassifier', rf_clf), ('ExtraTreesClassifier', et_clf),
                                        ('KNeighborsClassifier', knn_clf)], voting = 'hard')


ensembler_scores = cross_val_score(ensembler, X_train, y_train, cv = 10)
print(ensembler.__class__.__name__, ":::: Mean:", ensembler_scores.mean(), "Std Dev:", clf_scores.std())

'''

'\nfrom sklearn.tree import DecisionTreeClassifier\nfrom sklearn.ensemble import RandomForestClassifier, ExtraTreesClassifier, BaggingClassifier, VotingClassifier\nfrom sklearn.linear_model import LogisticRegression\nfrom sklearn.neighbors import KNeighborsClassifier\nfrom sklearn.svm import SVC\nfrom sklearn.model_selection import cross_val_score\n\nlr_clf = LogisticRegression() \ndt_clf = DecisionTreeClassifier() \nrf_clf = RandomForestClassifier() \net_clf = ExtraTreesClassifier() \nknn_clf = KNeighborsClassifier() \n\nclassifiers =[lr_clf, dt_clf, rf_clf, et_clf, knn_clf]\nfor clf in classifiers:\n    clf_scores = cross_val_score(clf, X_train, y_train, cv = 10)\n    bagging_clf = BaggingClassifier(clf, max_samples=0.4, max_features=10, random_state=0)\n    bagging_clf_scores = cross_val_score(bagging_clf, X_train, y_train, cv = 10)\n    print(clf.__class__.__name__, ":::: Mean:", clf_scores.mean(), ", Std Dev:", clf_scores.std())\n    print("Bagging", clf.__class__.__name__, ":::: 

In [51]:
'''
from sklearn.ensemble import BaggingClassifier
from sklearn.tree import DecisionTreeClassifier
from matplotlib.colors import ListedColormap
#Bagging 500 trees
bag_clf = BaggingClassifier(
    DecisionTreeClassifier(random_state=42), n_estimators=500,
    max_samples=100, oob_score=False, bootstrap=True, random_state=42)
bag_clf.fit(X_train, y_train)
y_pred = bag_clf.predict(X_test)
'''


'\nfrom sklearn.ensemble import BaggingClassifier\nfrom sklearn.tree import DecisionTreeClassifier\nfrom matplotlib.colors import ListedColormap\n#Bagging 500 trees\nbag_clf = BaggingClassifier(\n    DecisionTreeClassifier(random_state=42), n_estimators=500,\n    max_samples=100, oob_score=False, bootstrap=True, random_state=42)\nbag_clf.fit(X_train, y_train)\ny_pred = bag_clf.predict(X_test)\n'

In [52]:
'''
#Accuracy of Bagged Trees
from sklearn.metrics import accuracy_score
print('Accuracy of Bagged Trees: {}'.format(accuracy_score(y_test, y_pred)))
'''

"\n#Accuracy of Bagged Trees\nfrom sklearn.metrics import accuracy_score\nprint('Accuracy of Bagged Trees: {}'.format(accuracy_score(y_test, y_pred)))\n"

In [53]:
'''
tree_clf = DecisionTreeClassifier(random_state=42)
tree_clf.fit(X_train, y_train)
y_pred_tree = tree_clf.predict(X_test)
print('Accuracy of a Single Tree: {}'.format(accuracy_score(y_test, y_pred_tree)))
'''

"\ntree_clf = DecisionTreeClassifier(random_state=42)\ntree_clf.fit(X_train, y_train)\ny_pred_tree = tree_clf.predict(X_test)\nprint('Accuracy of a Single Tree: {}'.format(accuracy_score(y_test, y_pred_tree)))\n"

In [54]:
#bag_clf.predict([[1710018,0,47683.0,0,9217.5,0,8420.0,0,0,8929.0,26208.6,0,6673.3,0,52351.4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]])

In [55]:
'''
from sklearn.ensemble import BaggingClassifier
from sklearn.tree import DecisionTreeClassifier
from matplotlib.colors import ListedColormap
from sklearn.metrics import accuracy_score

#Bagging 500 trees
bag_clf = BaggingClassifier(
    DecisionTreeClassifier(random_state=42), n_estimators=500,
    max_samples=100, oob_score=True, bootstrap=True, random_state=42)
bag_clf.fit(X_train, y_train)
y_pred = bag_clf.predict(X_test)

print('OOB score is {}'.format(bag_clf.oob_score_))
print('Accuracy of Bagged Trees: {}'.format(accuracy_score(y_test, y_pred)))

'''

"\nfrom sklearn.ensemble import BaggingClassifier\nfrom sklearn.tree import DecisionTreeClassifier\nfrom matplotlib.colors import ListedColormap\nfrom sklearn.metrics import accuracy_score\n\n#Bagging 500 trees\nbag_clf = BaggingClassifier(\n    DecisionTreeClassifier(random_state=42), n_estimators=500,\n    max_samples=100, oob_score=True, bootstrap=True, random_state=42)\nbag_clf.fit(X_train, y_train)\ny_pred = bag_clf.predict(X_test)\n\nprint('OOB score is {}'.format(bag_clf.oob_score_))\nprint('Accuracy of Bagged Trees: {}'.format(accuracy_score(y_test, y_pred)))\n\n"

In [56]:
#The OOB decision function for each training istances is also available through oob_decision_function_ variable. In this case (since the base estimator
#DecisionTreeClassifier() has a predict_proba() method) the decision function return the class probabilities for each training instance.
#Let's print the first five observations.
#bag_clf.oob_decision_function_[0:5]

## Advanced Ensemble Techniques::Boosting

In [57]:
'''
from sklearn.ensemble import AdaBoostClassifier, GradientBoostingClassifier
from xgboost import XGBClassifier
from sklearn.model_selection import cross_val_score

ab_clf = AdaBoostClassifier() 
gb_clf = GradientBoostingClassifier() 
xgb_clf = XGBClassifier() 

ensembler = VotingClassifier(estimators=[('AdaBoostClassifier', ab_clf), ('GradientBoostingClassifier', gb_clf),
                                        ('XGBClassifier', xgb_clf)], voting = 'hard')

classifiers =[ab_clf, gb_clf, xgb_clf, ensembler]
for clf in classifiers:
    clf_scores = cross_val_score(clf, X_train, y_train, cv = 10)
    print(clf.__class__.__name__, ":::: Mean:", clf_scores.mean(), "Std Dev:", clf_scores.std())
    
'''

'\nfrom sklearn.ensemble import AdaBoostClassifier, GradientBoostingClassifier\nfrom xgboost import XGBClassifier\nfrom sklearn.model_selection import cross_val_score\n\nab_clf = AdaBoostClassifier() \ngb_clf = GradientBoostingClassifier() \nxgb_clf = XGBClassifier() \n\nensembler = VotingClassifier(estimators=[(\'AdaBoostClassifier\', ab_clf), (\'GradientBoostingClassifier\', gb_clf),\n                                        (\'XGBClassifier\', xgb_clf)], voting = \'hard\')\n\nclassifiers =[ab_clf, gb_clf, xgb_clf, ensembler]\nfor clf in classifiers:\n    clf_scores = cross_val_score(clf, X_train, y_train, cv = 10)\n    print(clf.__class__.__name__, ":::: Mean:", clf_scores.mean(), "Std Dev:", clf_scores.std())\n    \n'

### Advanced Ensemble Techniques:: Bagging meta-estimator

In [58]:
'''
from sklearn.ensemble import BaggingClassifier
from sklearn import tree
model = BaggingClassifier(tree.DecisionTreeClassifier(random_state=1))
model.fit(X_train, y_train)
model.score(X_test,y_test)
'''

'\nfrom sklearn.ensemble import BaggingClassifier\nfrom sklearn import tree\nmodel = BaggingClassifier(tree.DecisionTreeClassifier(random_state=1))\nmodel.fit(X_train, y_train)\nmodel.score(X_test,y_test)\n'

In [59]:
#model.predict([[1710018,0,47683.0,0,9217.5,0,8420.0,0,0,8929.0,26208.6,0,6673.3,0,52351.4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]])

### Advanced Ensemble Techniques::Random Forest

In [60]:
'''
from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier(max_depth=10, n_estimators=40)
rf.fit(X_train, y_train)
rf.score(X_test, y_test)
'''

'\nfrom sklearn.ensemble import RandomForestClassifier\n\nrf = RandomForestClassifier(max_depth=10, n_estimators=40)\nrf.fit(X_train, y_train)\nrf.score(X_test, y_test)\n'

In [61]:
'''
predict_train = rf.predict(X_train)
print('\nTarget on train data',predict_train)

# Accuray Score on train dataset
accuracy_train = accuracy_score(y_train,predict_train)
print('\n accuracy_score on train dataset :',accuracy_train)

# predict the target on the test dataset
predict_test = rf.predict(X_test)
print('\nTarget on test data:',predict_test)

# Accuracy Score on test dataset
accuracy_test = accuracy_score(y_test,predict_test)
print('\naccuracy_score on test dataset : ', accuracy_test)
'''

"\npredict_train = rf.predict(X_train)\nprint('\nTarget on train data',predict_train)\n\n# Accuray Score on train dataset\naccuracy_train = accuracy_score(y_train,predict_train)\nprint('\n accuracy_score on train dataset :',accuracy_train)\n\n# predict the target on the test dataset\npredict_test = rf.predict(X_test)\nprint('\nTarget on test data:',predict_test)\n\n# Accuracy Score on test dataset\naccuracy_test = accuracy_score(y_test,predict_test)\nprint('\naccuracy_score on test dataset : ', accuracy_test)\n"

In [62]:
#rf.predict([[1710018,0,47683.0,0,9217.5,0,8420.0,0,0,8929.0,26208.6,0,6673.3,0,52351.4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]])

#### OOB Evaluation in Random Forest

In [63]:
##https://mmuratarat.github.io/2019-08-06/ensemble-learning-details-examples
'''
from sklearn.metrics import accuracy_score

from sklearn.ensemble import RandomForestClassifier


rnd_clf = RandomForestClassifier(n_estimators=31, oob_score = True, random_state=42)
rnd_clf.fit(X_train, y_train)

y_pred_rf = rnd_clf.predict(X_test)

print('Accuracy of Random Forest algorithm: {}'.format(accuracy_score(y_test, y_pred_rf)))
print('OOB score is {}'.format(rnd_clf.oob_score_))

  '''                                                                    

"\nfrom sklearn.metrics import accuracy_score\n\nfrom sklearn.ensemble import RandomForestClassifier\n\n\nrnd_clf = RandomForestClassifier(n_estimators=31, oob_score = True, random_state=42)\nrnd_clf.fit(X_train, y_train)\n\ny_pred_rf = rnd_clf.predict(X_test)\n\nprint('Accuracy of Random Forest algorithm: {}'.format(accuracy_score(y_test, y_pred_rf)))\nprint('OOB score is {}'.format(rnd_clf.oob_score_))\n\n  "

In [64]:
#rnd_clf.predict([[1710018,0,47683.0,0,9217.5,0,8420.0,0,0,8929.0,26208.6,0,6673.3,0,52351.4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]])

In [65]:
#rnd_clf.estimators_

In [66]:
#fifth_small = rnd_clf.estimators_[5]
#fifth_small

In [67]:
# Create decision tree classifer object
'''
import matplotlib.pyplot as plt
clf = RandomForestClassifier(n_estimators=500, max_leaf_nodes = 16, random_state=0, n_jobs=-1)

# Train model
model = clf.fit(X, y)

# Calculate feature importances
importances = model.feature_importances_
'''

'\nimport matplotlib.pyplot as plt\nclf = RandomForestClassifier(n_estimators=500, max_leaf_nodes = 16, random_state=0, n_jobs=-1)\n\n# Train model\nmodel = clf.fit(X, y)\n\n# Calculate feature importances\nimportances = model.feature_importances_\n'

In [68]:
#y_pred_rf = model.predict(X_test)

#print('Accuracy of Random Forest algorithm: {}'.format(accuracy_score(y_test, y_pred_rf)))


### Advanced Ensemble Techniques::AdaBoost

In [69]:
'''
from sklearn.ensemble import AdaBoostClassifier
Ada_model = AdaBoostClassifier(random_state=1)
Ada_model.fit(X_train, y_train)
Ada_model.score(X_test,y_test)
'''

'\nfrom sklearn.ensemble import AdaBoostClassifier\nAda_model = AdaBoostClassifier(random_state=1)\nAda_model.fit(X_train, y_train)\nAda_model.score(X_test,y_test)\n'

In [70]:
#Ada_model.predict([[1710018,0,47683.0,0,9217.5,0,8420.0,0,0,8929.0,26208.6,0,6673.3,0,52351.4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]])

In [71]:
'''
from sklearn.ensemble import AdaBoostClassifier
Ada_model = AdaBoostClassifier(DecisionTreeClassifier(max_depth=100), n_estimators=200,
algorithm="SAMME.R", learning_rate=0.5)
Ada_model.fit(X_train, y_train)
Ada_model.score(X_test,y_test)
'''

'\nfrom sklearn.ensemble import AdaBoostClassifier\nAda_model = AdaBoostClassifier(DecisionTreeClassifier(max_depth=100), n_estimators=200,\nalgorithm="SAMME.R", learning_rate=0.5)\nAda_model.fit(X_train, y_train)\nAda_model.score(X_test,y_test)\n'

In [72]:
#Predict the response for test dataset
#y_pred_AdaBoost = Ada_model.predict(X_test)

#print('Accuracy of AdaBoost algorithm: {}'.format(accuracy_score(y_test, y_pred_AdaBoost)))

In [73]:
'''
predict_train = Ada_model.predict(X_train)
print('\nTarget on train data',predict_train)

# Accuray Score on train dataset
accuracy_train = accuracy_score(y_train,predict_train)
print('\n accuracy_score on train dataset :',accuracy_train)

# predict the target on the test dataset
predict_test = Ada_model.predict(X_test)
print('\nTarget on test data:',predict_test)

# Accuracy Score on test dataset
accuracy_test = accuracy_score(y_test,predict_test)
print('\naccuracy_score on test dataset : ', accuracy_test)
'''

"\npredict_train = Ada_model.predict(X_train)\nprint('\nTarget on train data',predict_train)\n\n# Accuray Score on train dataset\naccuracy_train = accuracy_score(y_train,predict_train)\nprint('\n accuracy_score on train dataset :',accuracy_train)\n\n# predict the target on the test dataset\npredict_test = Ada_model.predict(X_test)\nprint('\nTarget on test data:',predict_test)\n\n# Accuracy Score on test dataset\naccuracy_test = accuracy_score(y_test,predict_test)\nprint('\naccuracy_score on test dataset : ', accuracy_test)\n"

In [74]:
#Ada_model.predict([[1710018,0,47683.0,0,9217.5,0,8420.0,0,0,8929.0,26208.6,0,6673.3,0,52351.4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]])

### Advanced Ensemble Techniques::Gradient Boosting (GBM)

In [75]:
'''
from sklearn.ensemble import GradientBoostingClassifier
GBM_model= GradientBoostingClassifier(learning_rate=0.01,random_state=1)
GBM_model.fit(X_train, y_train)
GBM_model.score(X_test,y_test)




predict_train = GBM_model.predict(X_train)
print('\nTarget on train data',predict_train)

# Accuray Score on train dataset
accuracy_train = accuracy_score(y_train,predict_train)
print('\nAccuracy_score on train dataset :',accuracy_train)

# predict the target on the test dataset
predict_test = GBM_model.predict(X_test)
print('\nTarget on test data:',predict_test)

# Accuracy Score on test dataset
accuracy_test = accuracy_score(y_test,predict_test)
print('\naccuracy_score on test dataset : ', accuracy_test)
'''

"\nfrom sklearn.ensemble import GradientBoostingClassifier\nGBM_model= GradientBoostingClassifier(learning_rate=0.01,random_state=1)\nGBM_model.fit(X_train, y_train)\nGBM_model.score(X_test,y_test)\n\n\n\n\npredict_train = GBM_model.predict(X_train)\nprint('\nTarget on train data',predict_train)\n\n# Accuray Score on train dataset\naccuracy_train = accuracy_score(y_train,predict_train)\nprint('\nAccuracy_score on train dataset :',accuracy_train)\n\n# predict the target on the test dataset\npredict_test = GBM_model.predict(X_test)\nprint('\nTarget on test data:',predict_test)\n\n# Accuracy Score on test dataset\naccuracy_test = accuracy_score(y_test,predict_test)\nprint('\naccuracy_score on test dataset : ', accuracy_test)\n"

In [76]:
#GBM_model.predict([[1710018,0,47683.0,0,9217.5,0,8420.0,0,0,8929.0,26208.6,0,6673.3,0,52351.4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]])

### Advanced Ensemble Techniques::XGBoost

In [77]:
'''
import xgboost as xgb
xgb_model=xgb.XGBClassifier(random_state=1,learning_rate=0.01)
xgb_model.fit(X_train, y_train)
xgb_model.score(X_test,y_test)
'''

'\nimport xgboost as xgb\nxgb_model=xgb.XGBClassifier(random_state=1,learning_rate=0.01)\nxgb_model.fit(X_train, y_train)\nxgb_model.score(X_test,y_test)\n'

In [78]:
#xgb_model.predict([[1710018,0,47683.0,0,9217.5,0,8420.0,0,0,8929.0,26208.6,0,6673.3,0,52351.4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]])
#preds = xgb_model.predict(X_test)
#preds

### Advanced Ensemble Techniques::Light GBM

In [79]:
%%time
import lightgbm as lgb

from lightgbm import LGBMClassifier

lgbm = LGBMClassifier(objective='multiclass', random_state=5)
lgbm.fit(X, y)
y_pred = lgbm.predict(X_test)

Wall time: 23.2 s


In [80]:
lgbm.score(X_test,y_test)

0.43240793741652356

In [81]:
predict_train = lgbm.predict(X_train)
print('\nTarget on train data',predict_train)

# Accuray Score on train dataset
accuracy_train = accuracy_score(y_train,predict_train)
print('\nAccuracy_score on train dataset :',accuracy_train)

# predict the target on the test dataset
predict_test = lgbm.predict(X_test)
print('\nTarget on test data:',predict_test)

# Accuracy Score on test dataset
accuracy_test = accuracy_score(y_test,predict_test)
print('\naccuracy_score on test dataset : ', accuracy_test)


Target on train data ['OTH' 'OTH' 'OTH' ... 'MET' 'OTH' 'DM']

Accuracy_score on train dataset : 0.432203111910107

Target on test data: ['OTH' 'DM' 'OTH' ... 'DM' 'DM' 'OTH']

accuracy_score on test dataset :  0.43240793741652356


In [129]:
lgbm.predict([[1710018,0,47683.0,0,9217.5,0,8420.0,0,0,8929.0,26208.6,0,6673.3,0,52351.4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]])

array(['EQ'], dtype=object)

### Advanced Ensemble Techniques::CatBoost

In [ ]:
from catboost import CatBoostClassifier
from sklearn.datasets import make_classification
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import RepeatedStratifiedKFold
Cat=CatBoostClassifier(verbose=0, n_estimators=100)

cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=1)
n_scores = cross_val_score(Cat, X, y, scoring='accuracy', cv=cv, n_jobs=-1, error_score='raise')



Cat.fit(X,y)
Cat.score(x_test,y_test)

C:\Users\rahim\anaconda3\lib\site-packages\sklearn\model_selection\_split.py:672: UserWarning: The least populated class in y has only 5 members, which is less than n_splits=10.
  % (min_groups, self.n_splits)), UserWarning)


In [83]:


#from sklearn.datasets import make_multilabel_classification

# this will generate a random multi-label dataset
#X, y = make_multilabel_classification(sparse = True, n_labels = 20, allow_unlabeled = False)

In [124]:
from sklearn.model_selection import cross_val_score
from sklearn.tree import DecisionTreeClassifier
#clf = DecisionTreeClassifier(random_state=0)
#clf.fit(X,y)
#cross_val_score(clf, X, y, cv=10)

DecisionTreeClassifier(random_state=0)

In [125]:
#clf.score(X,y)

0.49741578956155474

In [126]:
#clf.predict([[1710018,0,47683.0,0,9217.5,0,8420.0,0,0,8929.0,26208.6,0,6673.3,0,52351.4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]])

array(['OTH'], dtype=object)

In [87]:
#len([1710018,0,47683.0,0,9217.5,0,8420.0,0,0,8929.0,26208.6,0,6673.3,0,52351.4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0])

In [88]:

'''
from sklearn.ensemble import RandomForestClassifier
#from sklearn.datasets import make_classification
#X, y = make_classification(n_samples=1000, n_features=31,
#                           n_informative=2, n_redundant=0,
#                           random_state=0, shuffle=False)
rf = RandomForestClassifier(max_depth=10, n_estimators=40)
rf.fit(X, y)
cross_val_score(rf, X, y, cv=10)
'''

'\nfrom sklearn.ensemble import RandomForestClassifier\n#from sklearn.datasets import make_classification\n#X, y = make_classification(n_samples=1000, n_features=31,\n#                           n_informative=2, n_redundant=0,\n#                           random_state=0, shuffle=False)\nrf = RandomForestClassifier(max_depth=10, n_estimators=40)\nrf.fit(X, y)\ncross_val_score(rf, X, y, cv=10)\n'

In [89]:
#rf.predict([[1701824, 0.0, 0.0, 0.0, 0.0, 0.0,0.0, 0.0, 4545.9, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 4545.9]])

In [90]:
#rf.score(X_test, y_test)

#### AdaBoostClassifier

In [91]:
'''
from sklearn.ensemble import AdaBoostClassifier
from sklearn.datasets import make_classification
#X, y = make_classification(n_samples=1000, n_features=31,
#                            n_informative=2, n_redundant=0,
#                            random_state=10, shuffle=False)
ada = AdaBoostClassifier(n_estimators=100, random_state=0)
ada.fit(X, y)
#AdaBoostClassifier(n_estimators=100, random_state=10)
'''



'\nfrom sklearn.ensemble import AdaBoostClassifier\nfrom sklearn.datasets import make_classification\n#X, y = make_classification(n_samples=1000, n_features=31,\n#                            n_informative=2, n_redundant=0,\n#                            random_state=10, shuffle=False)\nada = AdaBoostClassifier(n_estimators=100, random_state=0)\nada.fit(X, y)\n#AdaBoostClassifier(n_estimators=100, random_state=10)\n'

In [92]:
#ada.score(X, y)

In [93]:
#ada.predict([[1701194, 0.0, 0.0, 0.0, 0.0, 0.0,0.0, 0.0, 4545.9, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 4545.9]])

### KNeighborsClassifier

In [94]:
'''
from sklearn.neighbors import KNeighborsClassifier
neigh = KNeighborsClassifier(n_neighbors=2)
neigh.fit(X_train, y_train) 
neigh.score(X_train,y_train)
'''

'\nfrom sklearn.neighbors import KNeighborsClassifier\nneigh = KNeighborsClassifier(n_neighbors=2)\nneigh.fit(X_train, y_train) \nneigh.score(X_train,y_train)\n'

In [95]:
#neigh.predict([[1710018,0,47683.0,0,9217.5,0,8420.0,0,0,8929.0,26208.6,0,6673.3,0,52351.4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]])

In [96]:
#import sklearn.metrics as metrics

#prediction = neigh.predict(X_test)


In [97]:
#metrics.hamming_loss(y_test, prediction)

In [98]:
#metrics.accuracy_score(y_test, prediction)

In [99]:
#y_pred = neigh.predict(X_test)

In [100]:
#y_pred

In [101]:
#X_test

In [102]:
#y_pred=neigh.predict(X_test)
#df2=pd.DataFrame(data=y_test, columns=['y_test'])
#df2['y_pred']=y_pred
#df2

##### LogisticRegression

In [103]:
'''
#LogisticRegression
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression(solver='liblinear',multi_class='ovr' ,C=10)
lr.fit(X_train, y_train)
lr.score(X_test, y_test)
'''

"\n#LogisticRegression\nfrom sklearn.linear_model import LogisticRegression\nlr = LogisticRegression(solver='liblinear',multi_class='ovr' ,C=10)\nlr.fit(X_train, y_train)\nlr.score(X_test, y_test)\n"

In [104]:
#lr.predict([[1710018,0,47683.0,0,9217.5,0,8420.0,0,0,8929.0,26208.6,0,6673.3,0,52351.4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]])

In [105]:
#len([1710018,0,47683.0,0,9217.5,0,8420.0,0,0,8929.0,26208.6,0,6673.3,0,52351.4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0])

In [106]:
#len(["HEAT_ID", 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 124, 125, 126, 166, 180, 181, 182, 183, 190, 207, 208, 209, 210, 211, 212, 213, 214, 215, "Sum"])

In [107]:
#f=len([1710018,0,47683.0,0,9217.5,0,8420.0,0,0,8929.0,26208.6,0,6673.3,0,52351.4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0])
#f

#### RandomForestClassifier

In [108]:
'''
#Random Forest
from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier(max_depth=10, n_estimators=40)
rf.fit(X_train, y_train)
rf.score(X_test, y_test)
'''

'\n#Random Forest\nfrom sklearn.ensemble import RandomForestClassifier\n\nrf = RandomForestClassifier(max_depth=10, n_estimators=40)\nrf.fit(X_train, y_train)\nrf.score(X_test, y_test)\n'

In [109]:
#rf.predict([[1710018,0,47683.0,0,9217.5,0,8420.0,0,0,8929.0,26208.6,0,6673.3,0,52351.4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]])

### DecisionTreeClassifier

In [110]:
'''
#Decision tree
from sklearn import tree
DecTree = tree.DecisionTreeClassifier(max_depth=70, criterion='entropy',splitter='best')

DecTree.fit(X_train,y_train)
DecTree.score(X_train,y_train)
'''

"\n#Decision tree\nfrom sklearn import tree\nDecTree = tree.DecisionTreeClassifier(max_depth=70, criterion='entropy',splitter='best')\n\nDecTree.fit(X_train,y_train)\nDecTree.score(X_train,y_train)\n"

In [111]:
#DecTree.predict([[1710018,0,47683.0,0,9217.5,0,8420.0,0,0,8929.0,26208.6,0,6673.3,0,52351.4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]])

In [112]:
#DecTree.predict([[1701194, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 4545.9, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 4545.9]])

In [113]:
#ch=[1701194, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 4545.9, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 4545.9]
#len(ch)

### neural_network MLPClassifier

In [114]:
'''
from sklearn.neural_network import MLPClassifier
nralclf = MLPClassifier(solver='lbfgs', alpha=1e-5,hidden_layer_sizes=(5, 2), random_state=1)
nralclf.fit(X_train, y_train) 
nralclf.score(X_train,y_train)
'''

"\nfrom sklearn.neural_network import MLPClassifier\nnralclf = MLPClassifier(solver='lbfgs', alpha=1e-5,hidden_layer_sizes=(5, 2), random_state=1)\nnralclf.fit(X_train, y_train) \nnralclf.score(X_train,y_train)\n"

In [115]:
#nralclf.predict([[1710018,0,47683.0,0,9217.5,0,8420.0,0,0,8929.0,26208.6,0,6673.3,0,52351.4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]])

### LinearRegression

In [116]:

#from sklearn.linear_model import LinearRegression
#reg = LinearRegression()
#reg.fit(X_train, y_train) 
#reg.score(X_train,y_train)


In [117]:
#reg.predict([[1710018,0,47683.0,0,9217.5,0,8420.0,0,0,8929.0,26208.6,0,6673.3,0,52351.4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]])

### Support Vector Machine

In [118]:
#SVM
#from sklearn.svm import SVC
#svm = SVC(gamma='auto')
#svm.fit(X_train, y_train)
#svm.score(X_test, y_test)

In [119]:
#svm.predict([[1710018,0,47683.0,0,9217.5,0,8420.0,0,0,8929.0,26208.6,0,6673.3,0,52351.4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]])

## Pickle Model:

I choosed Decission tree Coz It has better score from others:

In [120]:
#import pickle
#Model save in file to write mode
#with open('finalized_model_neigh','wb') as f:
#    pickle.dump(neigh,f)

#### Select Model Appro:1

Creating test harness , we will use 10-fold cross -validation into 10 parts crane on the 9 part estimate to "accuracy"
split our data set into 10 part:train on the 9 part and test on the 1 part and this will be repeat for all combination of train and test pilots

In [121]:
#array=df.values # create a list of array with data
#X=array[:,0:29] # all the column from 0-4
#y=array[:,30] # the result/output cloumn 4
#validation_size=0.20 #test data 0.2% from the data set
#seed=6 #random number
#X_train, X_test, y_train, y_test = sklearn.model_selection.train_test_split(X,y,test_size=validation_size, random_state=seed)

In [122]:
#seed =6
#scoring='accuracy'

In [123]:
#spot Check Algorithms
#models=[]
#models.append(('LR',LogisticRegression()))
#models.append(('LDA',LinearDiscriminantAnalysis()))
#models.append(('KNN',KNeighborsClassifier()))
#models.append(('CART',DecisionTreeClassifier()))
#models.append(('NB',GaussianNB()))
#models.append(('SVM',SVC()))
#evalute each model in turn
#results=[]
#names=[]
#for name,model in models:
#    kfold=sklearn.model_selection.KFold(n_splits=10,random_state=seed)
#    cv_results=sklearn.model_selection.cross_val_score(model,X_train,y_train,cv=kfold,scoring=scoring)
#    results.append(cv_results)
#    names.append(name)
#    msg="%s: %f (%f)" % (name,cv_results.mean(),cv_results.std())
#    print(msg)

C:\Users\rahim\anaconda3\lib\site-packages\sklearn\model_selection\_split.py:297: FutureWarning: Setting a random_state has no effect since shuffle is False. This will raise an error in 0.24. You should leave random_state to its default (None), or set shuffle=True.
  FutureWarning
C:\Users\rahim\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
C:\Users\rahim\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
C:\Users\rahim\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
C:\Users\rahim\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
C:\Users\rahim\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
C:\Users\rahim\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
C:\Users\rahim\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
C:\Users\rahim\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
C:\Users\rahim\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
C:\Users\rahim\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
C:\Users\rahim\anaconda3\lib\site-packages\sklearn\model_selection\_split.py:297: FutureWarning: Setting a random_state has no effect since shuffle is False. This will raise an error in 0.24. You should leave random_state to its default (None), or set shuffle=True.
  FutureWarning
Result:LR: 0.129470 (0.003673)
Result: LDA: 0.106433 (0.004029)
C:\Users\rahim\anaconda3\lib\site-packages\sklearn\model_selection\_split.py:297: FutureWarning: Setting a random_state has no effect since shuffle is False. This will raise an error in 0.24. You should leave random_state to its default (None), or set shuffle=True.
  FutureWarning
Result: KNN: 0.010126 (0.000913)
C:\Users\rahim\anaconda3\lib\site-packages\sklearn\model_selection\_split.py:297: FutureWarning: Setting a random_state has no effect since shuffle is False. This will raise an error in 0.24. You should leave random_state to its default (None), or set shuffle=True.
  FutureWarning
Result: CART: 0.002114 (0.000815)
C:\Users\rahim\anaconda3\lib\site-packages\sklearn\model_selection\_split.py:297: FutureWarning: Setting a random_state has no effect since shuffle is False. This will raise an error in 0.24. You should leave random_state to its default (None), or set shuffle=True.
  FutureWarning
Result: NB: 0.002028 (0.003543)

### Approach 2: Use K Fold Cross validation

Manually try suppling models with different parameters to cross_val_score function with 5 fold cross validation